### PR TITLE
Basic writeable connection attributes + attribute permissions

### PR DESCRIPTION
**Pull request by noloerino**
This PR adds basic support for the GATT `WriteReq` (write w/ response) and `WriteCommand` (write w/o response) commands, permissions on attributes, and a demo that exposes a writeable attribute that toggles an LED. See also #146.

I tested the demo on a NRF52832DK board, and it seems to work as expected. No rush for reviewing this by the way; I don't need this feature any time soon, I just happened to have some spare time over the holidays.

Notes:
- This implementation lazily creates the attribute responsible for storing writeable data; it should also be possible to store an owned `Attribute` directly, but may be a little bit inelegant because of possibly differing sizes of the underlying byte buffer.
- Because the `group_end` function of `AttributeProvider` needs to return `Option<&Attribute>`, we can't directly return a lazily-generated attribute. The demo circumvents this by adding a static dummy attribute to end the characteristic and service groups.
- I replaced the `AttrValue` trait defined in #166 with `AsRef<[u8]>` because it became impossible for an `Attribute` to own its underlying data because the trait cannot be implemented for `[u8; N]`. I think this can be circumvented by passing a slice to a lazily produced `Attribute`. If anyone feels strongly enough about leaving the `AttrValue` trait in, then I can try to work around it.

The following related features are not yet implemented:
- Signed writes (`SignedWriteCommand`)
- Queued writes (`PrepareWriteReq` and `ExecuteWriteReq`; these would require some means of storing queued data that would take a little more design thinking)
- Callbacks on `write_attribute` for added flexibility (see #146)
- Returning an error if a `ReadReq` is made on an attribute that does not have readable permission (see comment in diff)

### CODE REVIEW

File: demos/nrf52-writeable-attributes/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    // use these fields.
+    led_pin: Pin<Output<PushPull>>,
+    led_buf: [u8; 1],
+}
+
+const PRIMARY_SERVICE_UUID16: Uuid16 = Uuid16(0x2800);
+const CHARACTERISTIC_UUID16: Uuid16 = Uuid16(0x2803);
+const GENERIC_ATTRIBUTE_UUID16: Uuid16 = Uuid16(0x1801);
+
+// TODO what UUID should this be? I took this from a course assignment :P
```
Comment (noloerino) [inline demos/nrf52-writeable-attributes/src/main.rs:53]: Any suggestions? I couldn't find any resources regarding how 128-bit UUIDs get determined.

File: rubble/src/att/server.rs [hunk 2]
```diff
@@ -197,10 +203,17 @@ impl<A: AttributeProvider> AttributeServer<A> {
                         self.attrs.for_attrs_in_range(
                             HandleRange::new(*handle, *handle),
                             |_provider, attr| {
-                                let value = if writer.space_left() < attr.value.as_slice().len() {
-                                    &attr.value.as_slice()[..writer.space_left()]
+                                // FIXME return if attribute is not readable
```
Comment (noloerino) [inline rubble/src/att/server.rs:206]: This callback can't return the proper error type with the current API design due (see comment).

File: demos/nrf52-writeable-attributes/src/main.rs (hunk 1 above)
Comment (jonas-schievink) [reply demos/nrf52-writeable-attributes/src/main.rs:55]: I think you're supposed to randomly generate one and stick to that

**Comment by jonas-schievink** [comment]
This looks pretty good so far, thanks!

One think I'd like to see is to get the new demo merged into the old one. That way there's less code to deal with, and it should be fine to have both writeable and readable attributes in that demo.

**Comment by noloerino** [comment]
Sure thing. Should I include both the LED and the battery service attributes?

**Comment by jonas-schievink** [comment]
Yeah that should be fine

File: demos/nrf52-demo/Cargo.toml [hunk 3]
```diff
@@ -19,6 +19,7 @@ cortex-m-rtic = "0.5.3"
 cortex-m-rt = "0.6.11"
 panic-semihosting = "0.5.3"
 bbqueue = "0.4.1"
+rtt-target = { version = "0.2.0", features = ["cortex-m"] }
```
Comment (jonas-schievink) [inline demos/nrf52-demo/Cargo.toml:22]: This seems unused

File: rubble/src/att/server.rs [hunk 4]
```diff
@@ -213,17 +226,36 @@ impl<A: AttributeProvider> AttributeServer<A> {
...
+                        .unwrap();
+                    Ok(())
+                } else {
+                    Err(AttError::new(ErrorCode::WriteNotPermitted, *handle))
+                }
+            }
+            AttPdu::WriteCommand { handle, value } => {
+                // WriteCommand shouldn't respond to the client even on failure
+                if self.attrs.attr_access_permissions(*handle).is_writeable() {
+                    self.attrs.write_attr(*handle, value.as_ref()).unwrap();
```
[the remark this answers is not in the document]
Comment (noloerino) [reply rubble/src/att/server.rs]: Fixed, and did the same for handling write requests as well. I noticed that in the original code, errors in handling `ReadReq` and `ExchangeMtuReq` are also unwrapped - should I change that as well while I'm at it?

[the remark this answers is not in the document]
Comment (jonas-schievink) [reply rubble/src/att/server.rs]: Not necessary for this PR, but feel free to file a followup PR if you want. I'll go ahead and merge this now, thanks!

**Pull request merged**